diff --git a/common/frame.c b/common/frame.c
index d21db7321d439d3163210a1cc57ae4f40b7132bf..50b29b84438608335c3aa46b5b1be45410a75c0f 100644
--- a/common/frame.c
+++ b/common/frame.c
@@ -405,6 +405,7 @@ int x264_frame_copy_picture( x264_t *h, x264_frame_t *dst, x264_picture_t *src )
     dst->opaque     = src->opaque;
     dst->mb_info    = h->param.analyse.b_mb_info ? src->prop.mb_info : NULL;
     dst->mb_info_free = h->param.analyse.b_mb_info ? src->prop.mb_info_free : NULL;
+    memcpy( dst->timecode, src->timecode, sizeof(dst->timecode) );
 
     uint8_t *pix[3];
     int stride[3];
